### ISSUE TITLE
test-link tests fail on subsequent runs

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#21637]
Against latest master: 4b89a4c7cbd3653518ecf743ed707f516d187e78

To reproduce, run the linker tests twice in a row. They fail the second time:

```
$ stage4/bin/zig build test-link -fqemu
$ stage4/bin/zig build test-link -fqemu
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-ld-script-path-error-aarch64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug aarch64-linux-gnu failure
error: 
========= should contain: ===============
error: missing library dependency: GNU ld script '/?/liba.so' requires 'libfoo.so', but file not found
========= but not found: ================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-ld-script-disallow-undefined-version-x86_64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-lld
         └─ zig build-lib add Debug x86_64-linux-gnu failure
error: 
========= should contain: ===============
error: ld.lld: version script assignment of 'ADD_1.0' to symbol 'sub' failed: symbol not defined
========= but not found: ================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-mismatched-cpu-architecture-error-x86_64-linux.4.19...6.10.3-musl-Debug-llvm-no-lld
         └─ zig build-exe main Debug x86_64-linux-musl failure
error: 
========= expected: =====================
invalid ELF machine type: AARCH64
note: while parsing /?/a.o

========= but found: ====================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-ld-script-path-error-x86_64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug x86_64-linux-gnu failure
error: 
========= should contain: ===============
error: missing library dependency: GNU ld script '/?/liba.so' requires 'libfoo.so', but file not found
========= but not found: ================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-ld-script-disallow-undefined-version-aarch64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-lld
         └─ zig build-lib add Debug aarch64-linux-gnu failure
error: 
========= should contain: ===============
error: ld.lld: version script assignment of 'ADD_1.0' to symbol 'sub' failed: symbol not defined
========= but not found: ================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-unresolved-error-aarch64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug aarch64-linux-gnu failure
error: 
========= expected: =====================
error: undefined symbol: foo
note: referenced by /?/a.o:.text.bar
note: referenced by /?/b.o:.text.main

========= but found: ====================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-unresolved-error-x86_64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug x86_64-linux-gnu failure
error: 
========= expected: =====================
error: undefined symbol: foo
note: referenced by /?/a.o:.text.bar
note: referenced by /?/b.o:.text.main

========= but found: ====================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-unknown-file-type-error-x86_64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug x86_64-linux-gnu failure
error: 
========= should contain: ===============
error: unexpected error: parsing input file failed with error InvalidLdScript
========= but not found: ================

=========================================
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-unknown-file-type-error-aarch64-linux.4.19...6.10.3-gnu.2.28-Debug-llvm-no-lld
         └─ zig build-exe main Debug aarch64-linux-gnu failure
error: 
========= should contain: ===============
error: unexpected error: parsing input file failed with error InvalidLdScript
========= but not found: ================

=========================================
^C571/2334] steps
├─ [3] zig build-exe test1 Debug x86_64-linux
```

**Comment by kubkon** [comment]
I'm having a look at that one and I'm not sure how to have the test build step try re-building the binary so that the errors are rethrown. Currently, it succeeds at first try since we actually try building the binary and catch compilation errors, however every subsequent re-build skips the compilation step altogether, or so it would appear. Not quite sure what to make of this yet.

**Comment by kubkon** [comment]
Taking one of the most recent bug reports as a minimal repro, here's what I see:

```zig
// build.zig
const std = @import("std");

pub fn build(b: *std.Build) void {
    const target = b.standardTargetOptions(.{});
    const optimize = b.standardOptimizeOption(.{});

    const step = b.step("test-link", "");

    const exe = b.addExecutable(.{
        .name = "main",
        .root_source_file = b.path("main.zig"),
        .target = target,
        .optimize = optimize,
    });
    exe.expect_errors = .{ .exact = &.{
        "error: undefined symbol: _objc_msgSend",
        "note: referenced implicitly",
    } };

    const bin_file = exe.getEmittedBin();
    bin_file.addStepDependencies(step);
}
```
```zig
pub fn main() !void {
    const msg_send_fn = @extern(
        *const fn () callconv(.C) usize,
        .{ .name = "objc_msgSend$initWithContentRect:styleMask:backing:defer:screen:" },
    );
    _ = @call(
        .auto,
        msg_send_fn,
        .{},
    );
}
```

* Using `zig build test-link` as expected succeeds first time round, fails subsequently:
```
$ zig build test-link
$ zig build test-link --verbose
/Users/kubkon/dev/zig/build/stage4/bin/zig build-exe -ODebug -Mroot=/Users/kubkon/dev/examples/issue-21637/src/main.zig --cache-dir /Users/kubkon/dev/examples/issue-21637/.zig-cache --global-cache-dir /Users/kubkon/.cache/zig --name main --zig-lib-dir /Users/kubkon/dev/zig/build/stage4/lib/zig/ --listen=- 
test-link
└─ zig build-exe main Debug native failure
error: 
========= expected: =====================
error: undefined symbol: _objc_msgSend
note: referenced implicitly

========= but found: ====================

=========================================
Build Summary: 0/2 steps succeeded; 1 failed
test-link transitive failure
└─ zig build-exe main Debug native failure
error: the following build command failed with exit code 1:
/Users/kubkon/dev/examples/issue-21637/.zig-cache/o/b34264ad2fe70e95070d961cc95c7032/build /Users/kubkon/dev/zig/build/stage4/bin/zig /Users/kubkon/dev/zig/build/stage4/lib/zig /Users/kubkon/dev/examples/issue-21637 /Users/kubkon/dev/examples/issue-21637/.zig-cache /Users/kubkon/.cache/zig --seed 0x9c6078fd -Z9c5ae44e8027db85 test-link --verbose
```

* Manually re-running `zig build-exe` command works fine tho:

```
$ /Users/kubkon/dev/zig/build/stage4/bin/zig build-exe -ODebug -Mroot=/Users/kubkon/dev/examples/issue-21637/src/main.zig --cache-dir /Users/kubkon/dev/examples/issue-21637/.zig-cache --global-cache-dir /Users/kubkon/.cache/zig --name main --zig-lib-dir /Users/kubkon/dev/zig/build/stage4/lib/zig/
error: undefined symbol: _objc_msgSend
    note: referenced implicitly
$ /Users/kubkon/dev/zig/build/stage4/bin/zig build-exe -ODebug -Mroot=/Users/kubkon/dev/examples/issue-21637/src/main.zig --cache-dir /Users/kubkon/dev/examples/issue-21637/.zig-cache --global-cache-dir /Users/kubkon/.cache/zig --name main --zig-lib-dir /Users/kubkon/dev/zig/build/stage4/lib/zig/
error: undefined symbol: _objc_msgSend
    note: referenced implicitly
```

**Comment by andrewrk** [comment]
Fix landed in 1340565e22e154db41e90a8d4f0b552176711ba0.
